coap/adapter.go

```diff
@@ -2,142 +2,172 @@
+	if err != nil {
+		return errors.Wrap(ErrUnauthorized, err)
+	}
 
-	val, ok := svc.obs[obsID]
-	if ok {
-		close(val.Cancel)
-		delete(svc.obs, obsID)
+	subject := fmt.Sprintf("%s.%s", chansPrefix, chanID)
+	if subtopic != "" {
+		subject = fmt.Sprintf("%s.%s", subject, subtopic)
 	}
-}
 
-// ListenResponses method handles ACK messages received from client.
-func (svc *adapterService) listenResponses(responses <-chan string) {
-	for {
-		id := <-responses
+	go func() {
+		<-c.Done()
+		svc.remove(subject, c.Token())
+	}()
 
-		val, ok := svc.get(id)
-		if ok {
-			val.StoreExpired(false)
-		}
+	obs, err := NewObserver(subject, c, svc.conn)
+	if err != nil {
+		c.Cancel()
+		return err
 	}
+	return svc.put(subject, c.Token(), obs)
 }
 
-func (svc *adapterService) Publish(msg messaging.Message) error {
-	return svc.ps.Publish(msg.Channel, msg)
-}
-
-func (svc *adapterService) Subscribe(chanID, subtopic, obsID string, o *Observer) error {
-	subject := chanID
+func (svc *adapterService) Unsubscribe(ctx context.Context, key, chanID, subtopic, token string) error {
+	ar := &mainflux.AccessByKeyReq{
+		Token:  key,
+		ChanID: chanID,
+	}
+	if _, err := svc.auth.CanAccessByKey(ctx, ar); err != nil {
+		return errors.Wrap(ErrUnauthorized, err)
+	}
+	subject := fmt.Sprintf("%s.%s", chansPrefix, chanID)
 	if subtopic != "" {
-		subject = fmt.Sprintf("%s.%s", chanID, subtopic)
+		subject = fmt.Sprintf("%s.%s", subject, subtopic)
 	}
 
-	err := svc.ps.Subscribe(subject, func(msg messaging.Message) error {
-		o.Messages <- msg
+	return svc.remove(subject, token)
+}
+
+func (svc *adapterService) put(endpoint, token string, o Observer) error {
+	svc.obsLock.Lock()
+	defer svc.obsLock.Unlock()
+
+	obs, ok := svc.observers[endpoint]
+	// If there are no observers, create map and assign it to the endpoint.
+	if !ok {
+		obs = observers{token: o}
+		svc.observers[endpoint] = obs
 		return nil
-	})
-	if err != nil {
-		return err
 	}
-
-	go func() {
-		<-o.Cancel
-		if err := svc.ps.Unsubscribe(subject); err != nil {
-			svc.log.Error(fmt.Sprintf("Failed to unsubscribe from %s.%s due to %s", chanID, subtopic, err))
+	// If observer exists, cancel subscription and replace it.
+	if sub, ok := obs[token]; ok {
+		if err := sub.Cancel(); err != nil {
+			return errors.Wrap(ErrUnsubscribe, err)
 		}
-	}()
-
-	// Put method removes Observer if already exists.
-	svc.put(obsID, o)
+	}
+	obs[token] = o
 	return nil
 }
 
-func (svc *adapterService) Unsubscribe(obsID string) {
-	svc.remove(obsID)
+func (svc *adapterService) remove(endpoint, token string) error {
+	svc.obsLock.Lock()
+	defer svc.obsLock.Unlock()
+
+	obs, ok := svc.observers[endpoint]
+	if !ok {
+		return nil
+	}
+	if current, ok := obs[token]; ok {
+		if err := current.Cancel(); err != nil {
+			return errors.Wrap(ErrUnsubscribe, err)
+		}
+	}
+	delete(obs, token)
+	// If there are no observers left for the endpint, remove the map.
+	if len(obs) == 0 {
+		delete(svc.observers, endpoint)
```

Comment:
do we need to remove this map? can it be reused here https://github.com/mainflux/mainflux/pull/1233/files#diff-a9745ffe405cfa25fb4f36761ce61eefR137

Reply:
Yes, we can do it, but it's a potential memory leak. If we don't remove it and nobody actually adds the observer for the corresponding topic, we'll likely end up flooded with empty maps.